### ISSUE TITLE
Allow passing a specific binary or example

### ISSUE THREAD

**Issue by yodaldevoid** [gnzlbg/cargo-asm#21]
Some projects have multiple binaries and many have examples. It would be nice to have `--bin` and `--example` flags.

**Comment by gnzlbg** [comment]
Thanks for filling the issue. 

I don't know if `rustc` can emit assembly for binaries (`--bin` or `--example`s, it does not really matter). I'll give this a try, maybe it is easy to support.

**Comment by yodaldevoid** [comment]
I suspect it can emit assembly for specific binaries as right now it appears to compile all binaries (`--bin` targets) in a crate and then only checks the assembly from the one produced from `src/bin.rs`.

**Comment by gnzlbg** [comment]
@yodaldevoid indeed! 

IIRC right now `cargo asm`/`cargo llvm-ir` scan the target directory for assembly / llvm-ir files of any name (it just introspect everything in the target directory having the right extension and assumes that it is part of the "binary"). I think that for this to reliably work for different binaries, examples, etc. `cargo asm` would need to restrict the search to files related to the binary/example requested. 

One problematic situation could be if you use a library in your binary, and query for the binary a path to a library function, which is probably going to get emitted in a different file (the one produced by compiling that library), so we somehow have to decide, for each binary / example, which assembly/llvm-ir files to take into account for the search, and which ones to ignore.

For example, if the same library is used in two different examples/binaries, but for example/binary it is compiled with different feature flags, producing different assembly, when you search for the assembly of a function of that library for the different examples/binaries, we would somehow need to pick the appropriate assembly/llvm-ir file for the library related to that example/binary, and ignore the other.

Maybe we can query `cargo` for the files related to a target/command inside the target directory, and scan only the relevant files. This would be the most robust way of implementing this, but I don't remember if `cargo metadata` already provides us with this information. If so, implementing this shouldn't be that hard (a matter of adding the `--bin` and `--example` options, and then just querying the relevant files using `cargo metadata` and using only those files when searching for the path requested). 

If `cargo metadata` does not output this information we'll have to file an issue in `cargo` upstream and fix that first, because I don't think we can reliably implement this here without that.

**Comment by velvia** [comment]
@gnzlbg  I would like to have disassembly for bench and examples.  Are you saying that in theory `cargo-asm` should be able to find methods in bench and examples?   I can't find mine, but perhaps that is because of some Cargo.toml settings or something.

**Comment by gnzlbg** [comment]
No, only that in theory `cargo-asm` could be extended to do that, but as things are right now, that wouldn't be trivial, and the work around isn't too hard (move the bench or example to a new crate with the right dependencies, and use `cargo-asm` on that).

**Comment by najamelan** [comment]
So do you have to specifically add a binary target to lib crates right now? When not running binary, almost my entire lib get's optimized away because it doesn't get used. `cargo asm` lists like one function out of 50... Most of the stuff is generic.
